### PR TITLE
Follow react rules of hooks

### LINKED ISSUE #1205: Distributor throws an error in WordPress 6.5 RC 3's new synced patterns editor
### Describe the bug

WordPress 6.5 is introducing a new flow for editing synced patterns. Instead of editing them in the current post/page, they're opened in a new editor window. This editor is throwing an error when Distributor is enabled. 

### Steps to Reproduce

1. Start on a test site running WP 6.5 RC 3 and Distributor. 
2. Edit a post/page, and create a new synced pattern.
3. Click on the synced pattern, and then click 'Edit original' in the block toolbar to open the synced pattern in the new editor.
4. Make edits and save them -- best I can tell they save fine!
5. Click the 'back' button to return to the post/page editor; note that the error displays there initially, too.

### Screenshots, screen recording, code snippet

New button to open the synced pattern editor:

image

The error that displays in the synced pattern editor: 

image

### Environment information

Device: Macbook, running OS 12.6.4
Browser: Firefox 123.0.1

### WordPress information

`
### wp-core ###

version: 6.5-RC3
site_language: en_US
user_language: en_US
timezone: America/Vancouver
permalink: /%category%/%postname%/
https_status: true
multisite: false
user_registration: 0
blog_public: 0
default_comment_status: undefined
environment_type: local
user_count: 56
dotorg_communication: true

### wp-paths-sizes ###

wordpress_path: /Users/laurelfulford/Local Sites/newspacktest/app/public
wordpress_size: loading...
uploads_path: /Users/laurelfulford/Local Sites/newspacktest/app/public/wp-content/uploads
uploads_size: loading...
themes_path: /Users/laurelfulford/Local Sites/newspacktest/app/public/wp-content/themes
themes_size: loading...
plugins_path: /Users/laurelfulford/Local Sites/newspacktest/app/public/wp-content/plugins
plugins_size: loading...
database_size: loading...
total_size: loading...

### wp-active-theme ###

name: Twenty Twenty-Four (twentytwentyfour)
version: 1.0
author: the WordPress team
author_website: https://wordpress.org
parent_theme: none
theme_features: core-block-patterns, post-thumbnails, responsive-embeds, editor-styles, html5, automatic-feed-links, widgets-block-editor, block-templates
theme_path: /Users/laurelfulford/Local Sites/newspacktest/app/public/wp-content/themes/twentytwentyfour
auto_update: Disabled

### wp-themes-inactive (8) ###

Newspack Joseph: version: 1.85.1, author: Automattic, Auto-updates disabled
Newspack Katharine: version: 1.85.1, author: Automattic, Auto-updates disabled
Newspack Nelson: version: 1.85.1, author: Automattic, Auto-updates disabled
Newspack Sacha: version: 1.85.1, author: Automattic, Auto-updates disabled
Newspack Scott: version: 1.85.1, author: Automattic, Auto-updates disabled
Newspack: version: 1.76.0, author: Automattic, Auto-updates disabled
Twenty Twenty-Three: version: 1.3, author: the WordPress team, Auto-updates disabled
Twenty Twenty-Two: version: 1.6, author: the WordPress team, Auto-updates disabled

### wp-plugins-active (1) ###

Distributor: version: 2.0.4, author: 10up Inc., Auto-updates disabled

### wp-plugins-inactive (43) ###

Akismet Anti-spam: Spam Protection: version: 5.3.1, author: Automattic - Anti-spam Team, Auto-updates disabled
Block Editor Colors: version: 1.2.4, author: MotoPress, Auto-updates disabled
Co-Authors Plus: version: 3.5.15, author: Mohammad Jangda, Daniel Bachhuber, Automattic, Auto-updates disabled
Debug Bar: version: 1.1.6, author: wordpressdotorg, Auto-updates disabled
Distributor: version: 2.0.3, author: 10up Inc., Auto-updates disabled
Hello Dolly: version: 1.7.2, author: Matt Mullenweg, Auto-updates disabled
Jetpack: version: 13.2.1, author: Automattic, Auto-updates disabled
Loco Translate: version: 2.6.7, author: Tim Whitlock, Auto-updates disabled
LottieFiles: version: 2.8.0, author: LottieFiles, Auto-updates disabled
Lottie Player - Block: version: 1.1.5, author: bPlugins LLC, Auto-updates disabled
Newspack: version: 3.4.0-alpha.3, author: Automattic, Auto-updates disabled
Newspack Ads: version: 1.46.0, author: Automattic, Auto-updates disabled
Newspack AP Date Format: version: 1.0.1, author: Automattic, Auto-updates disabled
Newspack Blocks: version: 3.0.5-alpha.1, author: Automattic, Auto-updates disabled
Newspack Campaigns: version: 2.31.0, author: Automattic, Auto-updates disabled
Newspack Listings: version: 2.14.1, author: Automattic, Auto-updates disabled
Newspack Manager: version: 1.45.0, author: Automattic, Auto-updates disabled
Newspack Media Partners: version: 1.1.0, author: Automattic, Auto-updates disabled
Newspack Multibranded Site: version: 1.2.0, author: Automattic, Auto-updates disabled
Newspack Network: version: 1.1.0, author: Automattic, Auto-updates disabled
Newspack Newsletters: version: 2.11.2, author: Automattic, Auto-updates disabled
Newspack Podcasts: version: 0.0.0, author: Automattic, Auto-updates disabled
Newspack Sponsors: version: 1.11.1, author: Automattic, Auto-updates disabled
Password Protected: version: 2.6.8, author: Password Protected, Auto-updates disabled
Pig Latin: version: 0.2, author: Nikolay Bachiyski, Auto-updates disabled
PublishPress Capabilities: version: 2.13.0, author: PublishPress, Auto-updates disabled
PWA: version: 0.8.0, author: PWA Plugin Contributors, Auto-updates disabled
Regenerate Thumbnails: version: 3.1.6, author: Alex Mills (Viper007Bond), Auto-updates disabled
Site Kit by Google: version: 1.122.0, author: Google, Auto-updates disabled
The Events Calendar: version: 6.3.4, author: The Events Calendar, Auto-updates disabled
Trust.txt Manager: version: 1.2, author: rtCamp, Auto-updates disabled
User Switching: version: 1.7.3, author: John Blackbourn & contributors, Auto-updates disabled
WooCommerce: version: 8.6.1, author: Automattic (latest version: 8.7.0), Auto-updates disabled
WooCommerce Memberships: version: 1.25.0, author: SkyVerge, Auto-updates disabled
WooCommerce Name Your Price: version: 3.5.2, author: Kathy Darling, Auto-updates disabled
WooCommerce Shipping & Tax: version: 2.5.3, author: WooCommerce, Auto-updates disabled
WooCommerce Stripe Gateway: version: 8.0.1, author: WooCommerce, Auto-updates disabled
WooCommerce Subscriptions: version: 5.0.1, author: WooCommerce, Auto-updates disabled
WordPress Beta Tester: version: 3.5.5, author: WordPress Upgrade/Install Team, Auto-updates disabled
WP Mail SMTP: version: 4.0.1, author: WP Mail SMTP, Auto-updates disabled
WP Rollback: version: 2.0.6, author: WP Rollback, Auto-updates disabled
Yoast SEO: version: 22.2, author: Team Yoast (latest version: 22.3), Auto-updates disabled
Yoast SEO Premium: version: 21.8, author: Team Yoast, Auto-updates disabled

### wp-media ###

image_editor: WP_Image_Editor_Imagick
imagick_module_version: 1809
imagemagick_version: ImageMagick 7.1.1-15 Q16 x86_64 98eceff6a:20230729 https://imagemagick.org
imagick_version: 3.7.0
file_uploads: 1
post_max_size: 1000M
upload_max_filesize: 300M
max_effective_size: 300 MB
max_file_uploads: 20
imagick_limits: 
	imagick::RESOURCETYPE_AREA: 128 GB
	imagick::RESOURCETYPE_DISK: 9.2233720368548E+18
	imagick::RESOURCETYPE_FILE: 6144
	imagick::RESOURCETYPE_MAP: 64 GB
	imagick::RESOURCETYPE_MEMORY: 32 GB
	imagick::RESOURCETYPE_THREAD: 1
	imagick::RESOURCETYPE_TIME: 9.2233720368548E+18
imagemagick_file_formats: 3FR, 3G2, 3GP, AAI, AI, APNG, ART, ARW, ASHLAR, AVI, AVS, BAYER, BAYERA, BGR, BGRA, BGRO, BMP, BMP2, BMP3, BRF, CAL, CALS, CANVAS, CAPTION, CIN, CIP, CLIP, CMYK, CMYKA, CR2, CR3, CRW, CUBE, CUR, CUT, DATA, DCM, DCR, DCRAW, DCX, DDS, DFONT, DNG, DOT, DPX, DXT1, DXT5, EPDF, EPI, EPS, EPS2, EPS3, EPSF, EPSI, EPT, EPT2, EPT3, ERF, EXR, FARBFELD, FAX, FF, FITS, FL32, FLV, FRACTAL, FTS, FTXT, G3, G4, GIF, GIF87, GRADIENT, GRAY, GRAYA, GROUP4, GV, HALD, HDR, HISTOGRAM, HRZ, HTM, HTML, ICB, ICO, ICON, IIQ, INFO, INLINE, IPL, ISOBRL, ISOBRL6, JNG, JNX, JPE, JPEG, JPG, JPS, JSON, JXL, K25, KDC, KERNEL, LABEL, M2V, M4V, MAC, MAP, MASK, MAT, MATTE, MEF, MIFF, MKV, MNG, MONO, MOV, MP4, MPC, MPEG, MPG, MPO, MRW, MSL, MSVG, MTV, MVG, NEF, NRW, NULL, ORA, ORF, OTB, OTF, PAL, PALM, PAM, PANGO, PATTERN, PBM, PCD, PCDS, PCL, PCT, PCX, PDB, PDF, PDFA, PEF, PES, PFA, PFB, PFM, PGM, PGX, PHM, PICON, PICT, PIX, PJPEG, PLASMA, PNG, PNG00, PNG24, PNG32, PNG48, PNG64, PNG8, PNM, POCKETMOD, PPM, PS, PS2, PS3, PSB, PSD, PTIF, PWP, QOI, RADIAL-GRADIENT, RAF, RAS, RAW, RGB, RGB565, RGBA, RGBO, RGF, RLA, RLE, RMF, RW2, SCR, SCT, SFW, SGI, SHTML, SIX, SIXEL, SPARSE-COLOR, SR2, SRF, STEGANO, STRIMG, SUN, SVG, SVGZ, TEXT, TGA, THUMBNAIL, TIFF, TIFF64, TILE, TIM, TM2, TTC, TTF, TXT, UBRL, UBRL6, UIL, UYVY, VDA, VICAR, VID, VIFF, VIPS, VST, WBMP, WEBM, WEBP, WMV, WPG, X3F, XBM, XC, XCF, XPM, XPS, XV, YAML, YCbCr, YCbCrA, YUV
gd_version: bundled (2.1.0 compatible)
gd_formats: GIF, JPEG, PNG, WebP, BMP
ghostscript_version: 9.56.1

### wp-server ###

server_architecture: Darwin 21.6.0 x86_64
httpd_software: nginx/1.16.0
php_version: 8.1.23 64bit
php_sapi: fpm-fcgi
max_input_variables: 4000
time_limit: 1200
memory_limit: 256M
max_input_time: 600
upload_max_filesize: 300M
php_post_max_size: 1000M
curl_version: 7.87.0 (SecureTransport) LibreSSL/3.3.6
suhosin: false
imagick_availability: true
pretty_permalinks: true
htaccess_extra_rules: false
current: 2024-03-19T19:46:38+00:00
utc-time: Tuesday, 19-Mar-24 19:46:38 UTC
server-time: 2024-03-19T12:46:34-07:00

### wp-database ###

extension: mysqli
server_version: 8.0.16
client_version: mysqlnd 8.1.23
max_allowed_packet: 16777216
max_connections: 151

### wp-constants ###

WP_HOME: undefined
WP_SITEURL: undefined
WP_CONTENT_DIR: /Users/laurelfulford/Local Sites/newspacktest/app/public/wp-content
WP_PLUGIN_DIR: /Users/laurelfulford/Local Sites/newspacktest/app/public/wp-content/plugins
WP_MEMORY_LIMIT: 40M
WP_MAX_MEMORY_LIMIT: 256M
WP_DEBUG: true
WP_DEBUG_DISPLAY: false
WP_DEBUG_LOG: false
SCRIPT_DEBUG: false
WP_CACHE: false
CONCATENATE_SCRIPTS: undefined
COMPRESS_SCRIPTS: undefined
COMPRESS_CSS: undefined
WP_ENVIRONMENT_TYPE: local
WP_DEVELOPMENT_MODE: undefined
DB_CHARSET: utf8mb4
DB_COLLATE: undefined

### wp-filesystem ###

wordpress: writable
wp-content: writable
uploads: writable
plugins: writable
themes: writable

### distributor ###

0: 2.0.4
1: Yes
2: laurel.fulford+distributor@automattic.com
3: 
	Override Author Byline: Yes
	Media Handling: Featured image only
4: N/A
5: 
	hub-site.local: 
- URL: http://hub-site.local/wp-json/
- Version: 2.0.2
- Status: valid
- Auth method: Username / Password
- Username: admin
- Roles Allowed to Push: administrator, editor
- Additional data: {"errors":[], "can_post":["post", "page", "attachment", "nav_menu_item", "wp_block", "wp_template", "wp_template_part", "wp_navigation", "product", "guest-author"], "can_get":["post", "page", "attachment", "nav_menu_item", "wp_block", "wp_template", "wp_template_part", "wp_navigation", "product", "guest-author"], "endpoint_suggestion":false, "is_authenticated":"yes"}
	wildly-crooked.jurassic.ninja: 
- URL: https://wildly-crooked.jurassic.ninja/wp-json/
- Version: N/A
- Status: valid
- Auth method: Username / Password
- Username: demo
- Roles Allowed to Push: administrator, editor
- Additional data: {"errors":[], "can_post":["post", "page", "attachment", "nav_menu_item", "wp_block", "wp_template", "wp_template_part", "wp_navigation", "newspack_nl_cpt"], "can_get":["post", "page", "attachment", "nav_menu_item", "wp_block", "wp_template", "wp_template_part", "wp_navigation", "newspack_nl_cpt"], "endpoint_suggestion":false, "is_authenticated":"yes"}

`

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by kirtangajjar**
### Description of the Change

Follow react rules of hooks

Closes #1205

### How to test the Change

1. Start on a test site running WP 6.5 RC 3 and Distributor. 
2. Edit a post/page, and create a new synced pattern.
3. Click on the synced pattern, and then click 'Edit original' in the block toolbar to open the synced pattern in the new editor.
4. Make edits and save them -- best I can tell they save fine!
5. Click the 'back' button to return to the post/page editor; note that the error displays there initially, too.

### Changelog Entry

> Fixed - JS synced patterns editor

### Credits

Props @kirtangajjar

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [x] All new and existing tests pass.

### CODE REVIEW

File: assets/js/gutenberg-plugin.js [hunk 1]
```diff
@@ -153,24 +153,34 @@ const DistributorIcon = () => (
...
 const DistributorPlugin = () => {
-	// Ensure the user has proper permissions
-	if (
-		dtGutenberg.noPermissions &&
-		1 === parseInt( dtGutenberg.noPermissions )
-	) {
-		return null;
-	}
-
 	// eslint-disable-next-line no-shadow, react-hooks/rules-of-hooks -- permission checks are needed.
```
Comment (peterwilsoncc) [inline assets/js/gutenberg-plugin.js]:
```suggestion
	// eslint-disable-next-line no-shadow
```

File: assets/js/gutenberg-plugin.js [hunk 2]
```diff
@@ -153,24 +153,34 @@ const DistributorIcon = () => (
...
 	const postType = useSelect( ( select ) =>
 		select( 'core/editor' ).getCurrentPostType()
 	);
 
-	// eslint-disable-next-line no-shadow, react-hooks/rules-of-hooks -- permission checks are needed.
+	// eslint-disable-next-line no-shadow -- permission checks are needed.
```
Comment (peterwilsoncc) [inline assets/js/gutenberg-plugin.js]:
```suggestion
	// eslint-disable-next-line no-shadow
```

File: assets/js/gutenberg-plugin.js [hunk 3]
```diff
@@ -153,24 +153,34 @@ const DistributorIcon = () => (
...
 	const postStatus = useSelect( ( select ) =>
 		select( 'core/editor' ).getCurrentPostAttribute( 'status' )
 	);
 
+	// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+	const distributorTopMenu = document.querySelector(
+		'#wp-admin-bar-distributor'
+	);
+
+	// eslint-disable-next-line no-shadow -- permission checks are needed.
```
Comment (peterwilsoncc) [inline assets/js/gutenberg-plugin.js]:
```suggestion
	// eslint-disable-next-line no-shadow
```

File: assets/js/gutenberg-plugin.js [hunk 4]
```diff
@@ -153,24 +153,34 @@ const DistributorIcon = () => (
...
+	const post = useSelect( ( select ) =>
+		select( 'core/editor' ).getCurrentPost()
+	);
+
+	// Ensure the user has proper permissions
```
Comment (peterwilsoncc) [inline assets/js/gutenberg-plugin.js]: May as well fix the comment style while we're changing this code.
```suggestion
	// Ensure the user has proper permissions.
```

**Review by peterwilsoncc — COMMENTED** [review]
This looks good, just a few nit picks to remove comments that are no longer accurate.

**Review by Sidsector9 — APPROVED** [review]
Tested and works well 👍

**Pull request merged**